src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -233,31 +176,82 @@ private unsafe SocketError ProcessIOCPResult(bool success, int bytesTransferred,
         ///     Note this buffer (if not empty) should already be pinned locally using `fixed` prior to the OS async call and until after this method returns.</param>
         /// <param name="cancellationToken">The cancellation token to use to cancel the operation.</param>
         /// <returns>The result status of the operation.</returns>
-        private unsafe SocketError ProcessIOCPResultWithDeferredAsyncHandling(bool success, int bytesTransferred, NativeOverlapped* overlapped, Memory<byte> bufferToPin, CancellationToken cancellationToken = default)
+        private unsafe SocketError ProcessIOCPResult(bool success, int bytesTransferred, ref NativeOverlapped* overlapped, Memory<byte> bufferToPin, CancellationToken cancellationToken)
         {
-            Debug.Assert(_asyncProcessingState == AsyncProcessingState.InProcess);
-
-            SocketError socketError = GetIOCPResult(success, overlapped);
+            SocketError socketError = GetIOCPResult(success, ref overlapped);
+            SocketFlags socketFlags = SocketFlags.None;
 
             if (socketError == SocketError.IOPending)
             {
-                RegisterToCancelPendingIO(overlapped, cancellationToken);
+                // Perform any required setup of the asynchronous operation.  Everything set up here needs to be undone in CompleteCore.CleanupIOCPResult.
+                if (cancellationToken.CanBeCanceled)
+                {
+                    Debug.Assert(_pendingOverlappedForCancellation == null);
+                    _pendingOverlappedForCancellation = overlapped;
+                    _registrationToCancelPendingIO = cancellationToken.UnsafeRegister(static s =>
```

Comment:
I wonder in what portion of the cases is the fast async completion so fast that it finishes before this cancellation registration, rendering it unnecessary.

Reply:
I can try to measure that on my repro and on ASP.NET Core's repro of the issue

Reply:
I hacked up our tests to track how many SAEA completions were sync, async, and async-masquerading-as-sync.  This isn't very scientific, as our tests execute very differently from what "real-world" code does, but as a data point, throughout the whole inner loop test suite on my machine this was only happened a few times:
```
Async as sync:
        DoOperationReceiveSingleBuffer: 6
        DoOperationConnectEx: 10
        DoOperationAccept: 1
        DoOperationSendPackets: 1
```

Regardless, based on https://github.com/dotnet/runtime/issues/61233, it would seem to happen "enough", at least in that particular usage.  It wouldn't take many such completions to trigger that behavior, as long as the launching of the operation ended up taking long enough to cause the spinning to fall back to a Thread.Sleep(1), at which point, things slow waaaaaayyyyy down.